rfc/002-fault-tolerance/README.md

```diff
@@ -0,0 +1,487 @@
+- `uuid`: a globally unique identifier for a partition.
+- `ACK`: a confirmation about an action.
+
+The semantics the function signatures are as follows:
+
+1. `get`: retrieves a list of `partition` objects corresponding one-to-one to
+   the requested list of partition UUIDs.
+
+2. `add`: registers a list of new partitions.
+
+3. `delete`: removes a list of partition from the metastore (doesn't delete the
+   underlying data)
+
+4. `replace`: performs an atomic deletion and addition of partition data, by
+   deleting the given UUIDs and adding new partition metadata.
+
+In theory, the functionality of `replace` is a superset of `add` and `delete`.
+For ease of exposition, we kept them as separate API functions.
+
+#### Subscription API
+
+The metastore must also expose a subscription mechanism to hook into the feed of
+updates, such that the catalog can learn about new partitions from other catalog
+instances.
+
+Once a catalog learns about a new partition, it will load its sketches so that
+they become available for querying.
+
+Given these requirements, we briefly looked at the following systems with regard
+to their suitability.
+
+#### Candidates
+
+##### FoundationDB
+
+FoundationDB (FDB) is a distributed key-value store which supports composing
+sets of read and write operations into atomic transactions. It also supports
+single-key watches.
+
+Unfortunately there is no way to hook into a feed of all committed transactions,
+so a layer for the CDC requirement must be implemented on top of the primitive
+operations. This could be implemented by writing the transaction log into a
+separate key space and keeping a dedicated key that counts the transactions to
+be watched by all participants. Whenever the watch triggers on a node it can
+consume the latest updates to synchronize the local representation of the state.
+The original key space that tracks the set of existing partitions would only
+serve to verify the validity of the transaction on request.
+
+To get a feel for the system, we briefly experimented with mapping our API to
+FDB primitives. We found the following:
+
+* Transactions are “all or nothing”, i.e., you can sequence multiple operations,
+  and if you commit and it goes through, they are all applied, otherwise rolled
+  back.
+
+* Multiple transactions that modify (read to or write from) the same key are not
+  forbidden, even if they overlap - to enable our use case, we’d have to verify
+  the existence of all the partitions we intend to delete, followed by doing all
+  the (SET, CLEAR) operations OR rollback / cancel the transaction.
+
+* While this strategy allows for serializing the transactions into a
+  conflict-free state, it does not help with each VAST node "seeing the full
+  picture", i.e., being aware of the transactions that are triggered from within
+  other VAST nodes.
+
+* There doesn’t seem to be a full CDC (change data capture) functionality, only
```

Comment:
Style confusion: Why use * here and - everywhere else for the list? (Not that it really matters...)